### ISSUE TITLE
Make ShardingSphere Proxy in GraalVM Native Image form available

### ISSUE THREAD

**Issue by linghengqian** [apache/shardingsphere#21347]
## Feature Request

**For English only**, other languages will not accept.

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot make decision by current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Is your feature request related to a problem?

- Yes. Refer to https://github.com/apache/shardingsphere/issues/21051 and https://github.com/apache/shardingsphere/issues/17779 .

### Describe the feature you would like.

- There is currently a nightly build of ShardingSphere Proxy at https://github.com/orgs/apache/packages/container/package/shardingsphere-proxy-native. But it didn't reach usable status because it wasn't ready yet. For some related error logs, please refer to the logs of https://github.com/apache/shardingsphere/actions/runs/3185102779/jobs/5194279014 .

- This issue aims to do the following to bring the nightly build to a usable state.

- [x]  Add nightly builds of GitHub Action Job and corresponding documentation.

- [x] Introduces `JUnit Platform` and `JUnit Vintage` to replace `JUnit 4` to provide the ability to unit test inside `GraalVM Native Image`. Reference https://github.com/graalvm/native-build-tools/issues/53 . This issue may need to wait for https://github.com/graalvm/native-build-tools/issues/313 to be closed before it is actually resolved.

- [x] Update most of the dependencies to match the metadata provided by https://github.com/oracle/graalvm-reachability-metadata , including but not limited to `com.zaxxer:HikariCP:3.4.2`. This also involves submitting metadata for new third-party dependencies to https://github.com/oracle/graalvm-reachability-metadata . Of course, this problem may also be alleviated by https://github.com/oracle/graalvm-reachability-metadata/issues/62 .

- [x] Introduce GraalVM `Agent` in `GraalVM Native Build Tools`. I hope to wait for https://github.com/graalvm/native-build-tools/issues/260 to be resolved. But if the problem mentioned in https://github.com/graalvm/native-build-tools/pull/315 is still not solved, I will follow what I mentioned in https://twitter.com/linghengqian/status/1575865141198934016?s=20&t=iVsRyWdOhmXeAMt_bnXRyg operation to introduce it.

- [x] Update the GraalVM CE version used by GitHub Actions to `22.3.0` or higher to resolve an issue involving `server.yaml` in `ShardingSphere Proxy` not being scanned from the classpath. Refer to https://github.com/oracle/graal/issues/5020 and https://github.com/oracle/graal/issues/5080 . 

- [x] Works with Groovy 4 specific issues. Its GraalVM Reachability Metadata seems very special that we shouldn't commit metadata to Oracle's Git repository. It's metadata looks like what is mentioned in https://github.com/apache/shardingsphere/pull/18175 and https://github.com/graalvm/native-build-tools/issues/324#issuecomment-1266655389 dynamically.

- [x] Working on a fix for Apache Calcite, due to https://github.com/apache/calcite/pull/2603 being closed, may need to be introduced in a unique way https://twitter.com/intjesus/status/1567276748257259520?s=20&t=nUlN7LI8e9_2KO9lPwCqaw, https://github.com/substrait-io/substrait/issues/21 and a component of https://github.com/substrait-io/substrait-java/tree/main/isthmus . This allowed us to finally get rid of https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/util/javac/JaninoCompiler.java like this, Janino related classes  .

- The PRs involved in this issue are as follows:
    - https://github.com/apache/shardingsphere/pull/17665
    - https://github.com/apache/shardingsphere/pull/18175 
    - https://github.com/apache/shardingsphere/pull/19737
    - https://github.com/apache/shardingsphere/pull/19964
    - https://github.com/apache/shardingsphere/pull/20937
    - https://github.com/apache/shardingsphere/pull/21046
    - https://github.com/apache/shardingsphere/pull/21109
    - https://github.com/apache/shardingsphere/pull/21341
    - https://github.com/apache/shardingsphere/pull/21571
    - https://github.com/apache/shardingsphere/pull/21657
    - https://github.com/apache/shardingsphere/pull/21688
    - https://github.com/apache/shardingsphere/pull/22169
    - https://github.com/oracle/graalvm-reachability-metadata/pull/121
    - https://github.com/apache/shardingsphere/pull/22495
    - https://github.com/oracle/graalvm-reachability-metadata/pull/167
    - https://github.com/oracle/graalvm-reachability-metadata/pull/168
    - https://github.com/oracle/graalvm-reachability-metadata/pull/169
    - https://github.com/oracle/graalvm-reachability-metadata/pull/170
    - https://github.com/oracle/graalvm-reachability-metadata/pull/174
    - https://github.com/oracle/graalvm-reachability-metadata/pull/176
    - https://github.com/oracle/graalvm-reachability-metadata/pull/193
    - https://github.com/apache/shardingsphere/pull/23728
    - https://github.com/apache/shardingsphere-elasticjob/pull/2179
    - https://github.com/apache/shardingsphere/pull/23873
    - https://github.com/apache/shardingsphere/pull/23957
    - https://github.com/oracle/graalvm-reachability-metadata/pull/218
    - https://github.com/apache/shardingsphere/pull/24177
    - https://github.com/apache/shardingsphere/pull/24271
    - https://github.com/apache/shardingsphere/pull/24912
    - https://github.com/apache/shardingsphere/pull/24929
    - https://github.com/apache/shardingsphere/pull/25181
    - https://github.com/apache/shardingsphere/pull/25509
    - https://github.com/apache/shardingsphere/pull/25515
    - https://github.com/apache/shardingsphere/pull/25523
    - https://github.com/apache/shardingsphere/pull/25632
    - https://github.com/apache/shardingsphere/pull/27487
    - https://github.com/apache/shardingsphere/pull/28060
    - https://github.com/apache/shardingsphere/pull/28194
    - https://github.com/apache/shardingsphere/pull/28213
    - https://github.com/apache/shardingsphere/pull/28340
    - https://github.com/oracle/graalvm-reachability-metadata/pull/380
    - https://github.com/oracle/graalvm-reachability-metadata/pull/388
    - https://github.com/graalvm/native-build-tools/pull/504
    - https://github.com/oracle/graalvm-reachability-metadata/pull/390
    - https://github.com/apache/shardingsphere/pull/28326
    - https://github.com/apache/shardingsphere/pull/28609
    - https://github.com/apache/shardingsphere/pull/28610
    - https://github.com/graalvm/native-build-tools/pull/484
    - https://github.com/graalvm/native-build-tools/pull/504
    - https://github.com/apache/shardingsphere/pull/29000
    - https://github.com/apache/shardingsphere/pull/29009
    - https://github.com/apache/shardingsphere/pull/29013
    - https://github.com/apache/shardingsphere/pull/29014
    - https://github.com/apache/shardingsphere/pull/29034

**Comment by linghengqian** [comment]
- I'm assuming the GraalVM reachability metadata about Apache Curator will be migrated out of Spring Cloud since their community just migrated the corresponding metadata for Consul Client 1.4.5.

**Comment by linghengqian** [comment]
- I did intentionally bump the version of Vertx used by the master branch of ShardingSphere from 4.3.3 to 4.3.4 to be faster to accept issues like https://github.com/vert-x3/issues/issues/604. But `vertx-sql-client:4.3.4` seems to have refactorings that https://vertx.io/blog/eclipse-vert-x-4-3-4/ does not directly document, such as around a series of changes to `io.vertx.mysqlclient.impl.protocol.ColumnDefinition`. I gave up on this idea for now.

- ![0b1392eae156581d107028bf2a08542](https://user-images.githubusercontent.com/20187731/195625596-d2ac0075-da45-4e5e-b706-e87ae42c9357.jpg)
- ![519b07ab568188cc3586ebacec99e2d](https://user-images.githubusercontent.com/20187731/195625642-06700d53-7320-45d6-8d2f-ace7bd802456.jpg)

- For `jetcd-core`, I explained it at https://github.com/apache/shardingsphere/pull/21552.

**Comment by linghengqian** [comment]
> * For `jetcd-core` and `atomikos`, I explained it at [Update library to serve building GraalVM Native Image #21552](https://github.com/apache/shardingsphere/pull/21552).

- I found and resolved the conflict around Jetcd and vertx in the new PR, refer to https://github.com/apache/shardingsphere/pull/21571 .

**Comment by linghengqian** [comment]
- I don't feel the need to switch to Junit Jupiter right away, so I still use Junit Vintage for the relevant PRs.
- ![image](https://user-images.githubusercontent.com/20187731/195968077-0a168cfb-4983-4f56-90df-9770d50f0ef7.png)

**Comment by linghengqian** [comment]
- The new properties of `HikarciCP` are blocking my PR and I feel like I need to handle it better in a separate PR. Because it also involves how to match to HikariCP `5.0.1`'s GraalVM reachability metadata.
- ![image](https://user-images.githubusercontent.com/20187731/195974067-1a13362b-f037-4563-ae41-ddc84a39521c.png)

**Comment by linghengqian** [comment]
- I will be adding a release profile to `distribution/proxy-native` to help with its release in ShardingSphere `5.2.2` and later.

**Comment by linghengqian** [comment]
- I saw some additional issues when I checked the GHA logs, maybe something needs to be done for https://github.com/netty/netty/issues/11369 which has not been closed.

**Comment by linghengqian** [comment]
- https://github.com/apache/shardingsphere/issues/20457 and https://github.com/apache/shardingsphere/pull/21599 finally allow ShardingSphere Proxy to be started without a separate `server.yaml`, which will be extremely and greatly simplifies the use of GraalVM Native Image.  But it still seems that the mode configuration cannot be added through DistSQL? 🤔

**Comment by linghengqian** [comment]
- The Log at https://github.com/apache/shardingsphere/actions/runs/3298856474/jobs/5441511861 shows that the CI of GraalVM Native Image is broken.  This question first came from https://github.com/apache/shardingsphere/pull/21677 .  I will submit a PR to fix it.

**Comment by linghengqian** [comment]
- I saw some incorrectly imported metadata at https://github.com/apache/shardingsphere/actions/runs/3303870293/jobs/5452309095 .  Looks like most of it is related to Netty.  I'm not quite sure what I need to do, considering that the Spring community actually has this problem.  Reference https://twitter.com/sdeleuze/status/1583493681386225664?t=2IBnsStmiaTmiqU_BqwpMw&s=19 .

- The dependencies of the metadata involved are as follows.
```groovy
implementation 'com.github.luben:zstd-jni:1.5.2-5'

implementation 'org.conscrypt:conscrypt-openjdk-uber:2.5.2'

implementation 'io.github.zawn:logback-lite:1.2.3'

implementation 'com.aayushatharva.brotli4j:native-windows-x86_64:1.8.0'

implementation 'com.ning:compress-lzf:1.1'

implementation 'io.netty:netty-codec:4.1.84.Final'

implementation 'net.jpountz.lz4:lz4:1.3.0'
```

**Comment by linghengqian** [comment]
> - It looks like https://github.com/oracle/graalvm-reachability-metadata/issues/50 explains the issues related to `Log4j2`.  Pending https://issues.apache.org/jira/browse/LOG4J2-2649 .  This seems to require us to force-pull the Log4j2 version in the native profile to `3.0.0-SNAPSHOT`.  Should I say using SNAPSHOT is a weird behavior?
> 
> - It just looks like log4j2's `3.0.0-SNAPSHOT` completely removes stuff related to `java.lang.invoke.LambdaMetafactory`, the graalvm team almost doesn't want to support such usage at all, which involves `SecurityManager`.  I assume that I should be able to find a simple operation in the native profile of the shardingsphere to specify the version of all the passed log4j2 dependencies to `3.0.0-SNAPSHOT`, and also remove the log4j1 from the upstream dependencies, and finally https://github.com/oracle/graalvm-reachability-metadata/issues/50 can be reopened and the metadata of log4j2 can be added. However, I am quite depressed how ShardingSphere CI Log can see a hamcrest related class when building GraalVM Native Image  warning, the test library is not in the final product.

- Hopefully https://github.com/oracle/graalvm-reachability-metadata/pull/108 will remove the warning related to Log4j2.

**Comment by linghengqian** [comment]
- In https://github.com/apache/shardingsphere/actions/runs/3481017089/jobs/5821561497 , the results displayed in the log are still not optimistic. I'm doing additional investigation locally.
```shell
Warning: Could not resolve class com.github.luben.zstd.ZstdCompressCtx for reflection configuration. Reason: java.lang.ClassNotFoundException: com.github.luben.zstd.ZstdCompressCtx.
Warning: Could not resolve class org.conscrypt.ConscryptEngine for reflection configuration. Reason: java.lang.ClassNotFoundException: org.conscrypt.ConscryptEngine.
Warning: Could not resolve class org.conscrypt.ConscryptEngine for reflection configuration. Reason: java.lang.ClassNotFoundException: org.conscrypt.ConscryptEngine.
Warning: Could not resolve class org.conscrypt.ConscryptEngine for reflection configuration. Reason: java.lang.ClassNotFoundException: org.conscrypt.ConscryptEngine.
Warning: Method ch.qos.logback.core.encoder.LayoutWrappingEncoder.setParent(Appender) not found.
Warning: Could not resolve class com.aayushatharva.brotli4j.Brotli4jLoader for reflection configuration. Reason: java.lang.ClassNotFoundException: com.aayushatharva.brotli4j.Brotli4jLoader.
Warning: Could not resolve class com.github.luben.zstd.Zstd for reflection configuration. Reason: java.lang.ClassNotFoundException: com.github.luben.zstd.Zstd.
Warning: Could not resolve class com.ning.compress.lzf.impl.UnsafeChunkDecoder for reflection configuration. Reason: java.lang.ClassNotFoundException: com.ning.compress.lzf.impl.UnsafeChunkDecoder.
Warning: Could not register io.netty.handler.codec.compression.Lz4FrameDecoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: net/jpountz/lz4/LZ4Exception.
Warning: Could not register io.netty.handler.codec.compression.Lz4FrameEncoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: net/jpountz/lz4/LZ4Exception.
Warning: Could not register io.netty.handler.codec.marshalling.CompatibleMarshallingDecoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteInput.
Warning: Could not register io.netty.handler.codec.marshalling.CompatibleMarshallingEncoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteOutput.
Warning: Could not register io.netty.handler.codec.marshalling.MarshallingDecoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteInput.
Warning: Could not register io.netty.handler.codec.marshalling.MarshallingEncoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteOutput.
Warning: Could not register method org.apache.commons.logging.impl.Log4JLogger.<init>(String) for reflection. Reason: java.lang.NoClassDefFoundError: org/apache/log4j/Category.
Warning: Could not resolve class org.apache.log4j.Level for reflection configuration. Reason: java.lang.ClassNotFoundException: org.apache.log4j.Level.
Warning: Could not resolve class org.apache.log4j.Priority for reflection configuration. Reason: java.lang.ClassNotFoundException: org.apache.log4j.Priority.
Warning: Could not resolve class org.apache.logging.log4j.Logger for reflection configuration. Reason: java.lang.ClassNotFoundException: org.apache.logging.log4j.Logger.
Warning: Could not resolve class org.apache.logging.log4j.core.LoggerContext for reflection configuration. Reason: java.lang.ClassNotFoundException: org.apache.logging.log4j.core.LoggerContext.
Warning: Could not resolve class org.apache.logging.log4j.message.DefaultFlowMessageFactory for reflection configuration. Reason: java.lang.ClassNotFoundException: org.apache.logging.log4j.message.DefaultFlowMessageFactory.
Warning: Could not resolve class org.apache.logging.log4j.message.ReusableMessageFactory for reflection configuration. Reason: java.lang.ClassNotFoundException: org.apache.logging.log4j.message.ReusableMessageFactory.
Warning: Could not resolve class org.conscrypt.Conscrypt for reflection configuration. Reason: java.lang.ClassNotFoundException: org.conscrypt.Conscrypt.
Warning: Could not resolve class org.conscrypt.OpenSSLContextImpl$TLSv13 for reflection configuration. Reason: java.lang.ClassNotFoundException: org.conscrypt.OpenSSLContextImpl$TLSv13.
Warning: Could not resolve class org.hamcrest.number.OrderingComparison for reflection configuration. Reason: java.lang.ClassNotFoundException: org.hamcrest.number.OrderingComparison.
Warning: Method com.zaxxer.hikari.HikariConfig.getScheduledExecutorService() not found.
Warning: Method com.zaxxer.hikari.HikariConfig.isInitializationFailFast() not found.
Warning: Method com.zaxxer.hikari.HikariConfig.isJdbc4ConnectionTest() not found.
Warning: Method com.zaxxer.hikari.HikariConfig.setInitializationFailFast(boolean) not found.
Warning: Method com.zaxxer.hikari.HikariConfig.setJdbc4ConnectionTest(boolean) not found.
Warning: Method com.zaxxer.hikari.HikariConfig.setScheduledExecutorService(ScheduledThreadPoolExecutor) not found.
Warning: Could not register io.netty.handler.codec.compression.Lz4FrameDecoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: net/jpountz/lz4/LZ4Exception.
Warning: Could not register io.netty.handler.codec.compression.Lz4FrameEncoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: net/jpountz/lz4/LZ4Exception.
Warning: Could not register io.netty.handler.codec.marshalling.CompatibleMarshallingDecoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteInput.
Warning: Could not register io.netty.handler.codec.marshalling.CompatibleMarshallingEncoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteOutput.
Warning: Could not register io.netty.handler.codec.marshalling.MarshallingDecoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteInput.
Warning: Could not register io.netty.handler.codec.marshalling.MarshallingEncoder: queryAllPublicMethods for reflection. Reason: java.lang.NoClassDefFoundError: org/jboss/marshalling/ByteOutput.
```

**Comment by linghengqian** [comment]
- When ShardingSphere tests the building of GraalVM Native Image, it is found that `io.vertx.core.impl.AddressResolver` lacks GraalVM accessibility metadata. This class belongs to `io.vertx:vertx-core:4.3.3`. Therefore, it is linked to the new issue. For example https://github.com/netty/netty/issues/11369 .
- ![image](https://user-images.githubusercontent.com/20187731/202256280-606ecb58-b6a2-4076-9cad-ea6a6327eb94.png)

**Comment by linghengqian** [comment]
- I initialized the first part of GRM for shardingsphere-jdbc-core at https://github.com/linghengqian/graalvm-trace-metadata-smoketest/issues/1. But I saw https://github.com/apache/shardingsphere/pull/20255 related issue. I will submit a PR for this. I also opened https://github.com/oracle/graal/issues/5522.

**Comment by linghengqian** [comment]
- Once the https://github.com/apache/shardingsphere/pull/22982 is merged, nativeTest with vertx error log will no longer exist.

**Comment by linghengqian** [comment]
- Blocked by https://github.com/apache/shardingsphere/issues/22899 .

**Comment by zhfeng** [comment]
Hi @linghengqian 

I'm just wondering what version of GraalVM you want to target for shardingsphere-proxy-native? I'm trying to run `mvn clean install -Pnative` in `sharding-sphere/distribution/proxy-native`, but it is failing wiht
```
'gu' tool wasn't found. This probably means that JDK at isn't a GraalVM distribution.
```

It seems I need to install the GraalVM locally?

**Comment by linghengqian** [comment]
@zhfeng 
- You of course need to configure GraalVM's toolchain locally, and `mvn clean install -Pnative` doesn't make sense since I only defined the `package` task. What you need to do in `Linux` is somewhat similar to what I pointed out at https://shardingsphere.apache.org/document/5.3.0/en/user-manual/shardingsphere-proxy/startup/graalvm-native-image/ .
```shell
sudo apt install unzip zip curl sed -y
curl -s "https://get.sdkman.io" | bash
source "$HOME/.sdkman/bin/sdkman-init.sh"
sdk install java 22.3.r17-grl
sdk use java 22.3.r17-grl
gu install native-image
sudo apt-get install build-essential libz-dev zlib1g-dev -y

git clone git@github.com:apache/shardingsphere.git
cd ./shardingsphere/
./mvnw -am -pl distribution/proxy-native -B -Pnative -DskipTests -Dmaven.javadoc.skip=true -Dcheckstyle.skip=true -Dspotless.apply.skip=true -Drat.skip=true clean package
```
- I need to point out that the generated product is not available, because I was blocked by a series of issues I mentioned above, I first need to expose the SPI of ShardingSphere ActualDataNodes. The reason for the successful build is that I set the `--report-unsupported-elements-at-runtime` build args.

- `nativeTest` in `maven-plugin` of `graalvm native build tools` is not actually available, need to wait for https://github.com/graalvm/native-build-tools/issues/260 to close. So I actually test the graalvm reachability metadata generated by shardingsphere at https://github.com/linghengqian/graalvm-trace-metadata-smoketest because `gradle-plugin` has agent mode.
```shell
sdk install java 22.3.r17-grl
sdk use java 22.3.r17-grl
gu install native-image
sudo apt-get install build-essential libz-dev zlib1g-dev -y

git clone git@github.com:linghengqian/graalvm-trace-metadata-smoketest.git
cd ./graalvm-trace-metadata-smoketest/shardingsphere-jdbc-core
./gradlew -Pagent clean test
./gradlew metadataCopy --task test
./gradlew clean nativeTest
```
- This way, you can see why I need to avoid the use of GroovyShell for ShardingSphere ActualDataNodes. Refer to https://github.com/oracle/graal/issues/5522 and internal discussion in the Groovy and GraalVM Slack group.
- By the way, ShardingSphere's CI actually uses GraalVM 23.0 Dev, because there are some bugs in GraalVM CE 22.3, if you want to explore, you can pay attention to the historical PR at https://github.com/oracle/graalvm-reachability-metadata.
- Make sure your build machine has enough memory, at least **10GB** of build memory.
- ![image](https://user-images.githubusercontent.com/20187731/211149342-664ddacf-8615-41cb-bb06-da4d5c2a0fb1.png)
- ![image](https://user-images.githubusercontent.com/20187731/211149376-31acbfc7-f5a9-43b4-818d-a98bb81a24c3.png)

**Comment by zhfeng** [comment]
Thanks @linghengqian and if I understand correctly, Groovy rules can not be supported in shardingsphere-proxy-native right now? and what is the current status? is it possible to run proxy-native with mysql backends?

**Comment by zhfeng** [comment]
I asked because I met the similar problems during make `quarkus-shardingsphere-jdbc` to work in native mode. I had to exclude some dependencies and gave up with some groovy issues finally. So I think shardingsphere could have a alternative mothed to define the rules, is it right?

**Comment by linghengqian** [comment]
@zhfeng 

- I need to clarify something, not that Groovy does not support GraalVM NativeTest, but components like GroovyShell, Calcite, JRuby, Log4j2 2.x are not supported in GraalVM Native Image, because APIs such as Invokedynamic or Runtime bytecode generation are involved.

- Once I resolve https://github.com/apache/shardingsphere/issues/22899 , we can bypass GroovyShell, such as generated `java.util.List` of ActualDataNodes through pure Java API. We could also consider using Truffle Javascript or Truffle Python's SPI implementation to generate ActualDataNodes . 

- We can even consider introducing the SPI implementation of Truffle Espresso to put the operation of GroovyShell into Truffle Espresso. Truffle Espresso **JITs** GroovyShell operations to the guest JVM process within GraalVM Native Image. Of course, there are limitations, such as calling methods on the Host JVM process will become troublesome. And the JIT mode of Truffle Espresso can only run under Linux, although in GraalVM Native Image, Truffle Espresso can run under Linux, Windows, and Macos.

- Although ShardingSphere Proxy Native is currently unavailable, I can start testing around ShardingSphere Proxy Native once I solve a series of problems surrounding this issue and run through the `nativeTest` of `graalvm native build tools`. I wish I could fix related issues sooner, but the GraalVM ecosystem has always had issues that I need to focus on.

**Comment by zhfeng** [comment]
Yeah, thanks for the explanation.

The idea for `introducing the SPI implementation of Truffle Espresso to put the operation of GroovyShell into Truffle Espresso` is very impressed.

- So https://github.com/apache/shardingsphere/issues/22899 could be also helpful to `quarkus-shardingsphere-jdbc`
- Is there any example to define rules through pure Java API in yaml configuation file? or we have to do it programmatically?

**Comment by linghengqian** [comment]
@zhfeng 

- If you are referring to pure Java API to write ShardingSphere configuration, this is not a problem, but it is a problem for ActualDataNodes, which is a string. In the current version, the string of ActualDataNodes will be converted into a `java.util.List` containing all real tables through GroovyShell on the backend of ShardingSphere, and there is currently no way to intervene.

**Comment by zhfeng** [comment]
@linghengqian OK, I understand and looking forward to see https://github.com/apache/shardingsphere/issues/22899 to be resolved in the next release.

**Comment by linghengqian** [comment]
- Due to my lack of knowledge about this part of the new SPI, I temporarily put https://github.com/apache/shardingsphere/issues/22899 on hold. 
- But Truffle Espresso was introduced at https://github.com/apache/shardingsphere/pull/23873. 
- Early contextual discussions are on the GraalVM Slack group. Espresso has a lot of limitations compared to other Truffle API implementation languages, so I initially verified the contents of this PR in a Gradle project, if anyone has experience with GraalVM Truffle, welcome to review.

**Comment by linghengqian** [comment]
- After the introduction of Truffle Espresso, due to the need to download additional GraalVM CE and Truffle Espresso components, the Docker Image of ShardingSphere Proxy Native has expanded to 1.1GB.  I opened https://github.com/graalvm/graalvm-jdk-downloader/issues/5 to track it.

**Comment by linghengqian** [comment]
- At https://github.com/oracle/graalvm-reachability-metadata/pull/215, `dnestoro` mentions that he is refactoring the entire `org.graalvm.internal.tck` gradle plugin to address https://github.com/oracle/graalvm-reachability-metadata/issues/4, all PRs are broken for review. 

- For the remaining unsubmitted dependent GraalVM reachability metadata, I will directly open a PR on the ShardingSphere side for temporary storage, which is expected to involve GraalVM reachability metadata entries of  `20` maven libraries.

**Comment by linghengqian** [comment]
> - Since the `org.graalvm.truffle:truffle-api:23.0.1` with GraalVM CE 23.0.0 Release broke the nightly build of ShardingSphere Proxy Native, I have moved Truffle Espresso out of the final product in https://github.com/apache/shardingsphere/pull/27487 . 
> 
> - The new module `shardingsphere-infra-expr-purelist` means that the `Row Value Expressions` feature of ShardingSphere Proxy Native will not be able to parse the Groovy syntax, but will only use plain lists, and the documentation will be updated synchronously. 
> 
> - Issues with GraalVM are tracked in places like https://github.com/oracle/graal/issues/7072 .

- I'm not yet sure about the impact of https://github.com/oracle/graal/issues/6852#issuecomment-1667594171 on the ShardingSphere side.  Because Espresso is obviously a jar of GPL LICENSE, and this series of jars released following GraalVM CE 23.1 are all compiled with JDK17. And ShardingSphere is using a low version truffle-api compiled by JDK8 through different Maven Profiles.

**Comment by linghengqian** [comment]
- Now ShardingSphere's Native Image Build CI is starting to fail miraculously after https://github.com/apache/shardingsphere/pull/28099. Refer to https://github.com/apache/shardingsphere/actions/runs/5912312066/job/16035597496 . At least https://github.com/apache/shardingsphere/commit/fd574b8148331d9d2d23a2843ef83fef07323018 is normal.
```shell
[1/8] Initializing...                                                                                   (19.9s @ 0.39GB)
 Java version: 17.0.8+7, vendor version: GraalVM CE 17.0.8+7.1
 Graal compiler: optimization level: 2, target machine: x86-64-v3
 C compiler: gcc (linux, x86_64, 11.4.0)
 Garbage collector: Serial GC (max heap size: 80% of RAM)
 2 user-specific feature(s)
 - com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature
 - com.oracle.svm.thirdparty.gson.GsonFeature
[2/8] Performing analysis...  []                                                                        (56.0s @ 1.20GB)
  11,564 (82.58%) of 14,003 types reachable
  18,916 (60.69%) of 31,169 fields reachable
  58,086 (53.95%) of 107,671 methods reachable
   3,818 types,   148 fields, and 2,673 methods registered for reflection

Error: java.util.concurrent.ExecutionException: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: No instances of ch.qos.logback.core.status.InfoStatus are allowed in the image heap as this class should be initialized at image runtime. Object has been initialized by the io.grpc.netty.shaded.io.netty.channel.AbstractChannel class initializer with a trace: 
        at ch.qos.logback.core.status.InfoStatus.<init>(InfoStatus.java:18)
        at ch.qos.logback.classic.util.ContextInitializer.statusOnResourceSearch(ContextInitializer.java:156)
        at ch.qos.logback.classic.util.ContextInitializer.getResource(ContextInitializer.java:125)
        at ch.qos.logback.classic.util.ContextInitializer.findURLOfDefaultConfigurationFile(ContextInitializer.java:114)
        at ch.qos.logback.classic.util.ContextInitializer.autoConfig(ContextInitializer.java:132)
        at org.slf4j.impl.StaticLoggerBinder.init(StaticLoggerBinder.java:84)
        at org.slf4j.impl.StaticLoggerBinder.<clinit>(StaticLoggerBinder.java:55)
        at org.slf4j.LoggerFactory.bind(LoggerFactory.java:150)
        at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:124)
        at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:417)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.Slf4JLoggerFactory.<init>(Slf4JLoggerFactory.java:42)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.Slf4JLoggerFactory$NopInstanceHolder.<clinit>(Slf4JLoggerFactory.java:63)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.Slf4JLoggerFactory.getInstanceWithNopCheck(Slf4JLoggerFactory.java:59)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.InternalLoggerFactory.useSlf4JLoggerFactory(InternalLoggerFactory.java:62)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.InternalLoggerFactory.newDefaultFactory(InternalLoggerFactory.java:42)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.InternalLoggerFactory.getDefaultFactory(InternalLoggerFactory.java:111)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.InternalLoggerFactory.getInstance(InternalLoggerFactory.java:134)
        at io.grpc.netty.shaded.io.netty.util.internal.logging.InternalLoggerFactory.getInstance(InternalLoggerFactory.java:127)
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannel.<clinit>(AbstractChannel.java:45)
.  To fix the issue mark ch.qos.logback.core.status.InfoStatus for build-time initialization with --initialize-at-build-time=ch.qos.logback.core.status.InfoStatus or use the the information from the trace to find the culprit and --initialize-at-run-time=<culprit> to prevent its instantiation.

Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception
------------------------------------------------------------------------------------------------------------------------
                       11.8s (15.1% of total time) in 68 GCs | Peak RSS: 3.08GB | CPU load: 4.74
========================================================================================================================
Finished generating 'apache-shardingsphere-proxy-native' in 1m 16s.
com.oracle.svm.driver.NativeImage$NativeImageError
        at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.showError(NativeImage.java:1982)
        at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.build(NativeImage.java:1598)
        at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.performBuild(NativeImage.java:1557)
        at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.main(NativeImage.java:1531)

```
- The reason for this CI failure is simple. The error happens because `org.apache.shardingsphere:shardingsphere-proxy-bootstrap` has dependency on `io.grpc:grpc-netty-shaded:1.51.0`. In the shared metadata repository there is no metadata for `io.grpc:grpc-netty-shaded:1.51.0` that would override metadata from `io.grpc:grpc-netty-shaded:1.51.0` library like there is for `io.netty` libs. The early investigation is located in https://github.com/micronaut-projects/micronaut-gcp/issues/532 . It looks like the GraalVM Team hasn't reached a result yet. 
- I'll open a PR to temporarily disable `org.apache.shardingsphere:shardingsphere-cluster-mode-repository-etcd` into the Native Image build process.

**Comment by linghengqian** [comment]
- At https://github.com/oracle/graal/pull/7277, the memory usage behavior of GraalVM CE Dev 23.1.0 for JDK21 has changed again.  After that one should consider rolling back the changes made by https://github.com/apache/shardingsphere/pull/28213.  

- At the same time, we still need to wait for the release of Lombok to solve https://github.com/apache/shardingsphere/issues/28293.  

- The URL used by the documentation should be changed, as the GraalVM Team provided a new documentation URL Tag a while ago at https://github.com/quarkusio/quarkus/issues/34929 .

- Once the issue involved in https://github.com/apache/shardingsphere/pull/28340 is resolved, we will hopefully open nativeTest on the ShardingSphere side, although the support for multi-module projects in the Maven Plugin of Native Build Tools 0.9.25 still exists Visible problems.

**Comment by RaigorJiang** [comment]
Great work 👍

**Comment by zhfeng** [comment]
Thanks @linghengqian and it is really very nice work! Do you have any plan to write a simple application to compare the performance between the `GraalVM Native` and `JVM`? I also write a [quarkus extension](https://github.com/quarkiverse/quarkus-shardingsphere-jdbc/tree/1.x) to run with `shardingsphere-jdbc` in native mode which compile and evaluate the `groovy` rules at the build time.

**Comment by linghengqian** [comment]
> Thanks @linghengqian and it is really very nice work! Do you have any plan to write a simple application to compare the performance between the `GraalVM Native` and `JVM`? I also write a [quarkus extension](https://github.com/quarkiverse/quarkus-shardingsphere-jdbc/tree/1.x) to run with `shardingsphere-jdbc` in native mode which compile and evaluate the `groovy` rules at the build time.

- Sorry, it is not convenient for me to write a unit test for performance comparison because I lack the knowledge related to stress testing ShardingSphere.  

- For groovyshell, the unit tests related to nativeTest have skipped the use of groovyshell. Please refer to https://shardingsphere.apache.org/document/current/cn/user-manual/shardingsphere-jdbc/graalvm-native-image/ .  The relevant unit tests are located at `org.apache.shardingsphere:shardingsphere-infra-nativetest`.

- This involves the buildArgs of `--report-unsupported-elements-at-runtim` which downstream consumers must use.

- To facilitate review, I compressed the size of the initial subset of unit tests.  What can be verified now is that in the `standalone mode` of ShardingSphere, the five functions and derivative functions of `sharding`, `readwrite-splitting`, `mask`, `shadow` and `encrypt` can be used under native image.  For `cluster mode` and `DistSQL`, I have not added corresponding verification yet, and additional GraalVM Reachability Metadata json files should be required.  

- Since the modifications to these json files are not actually related to the current issue, I will discuss them in a new issue later.

- The core problem solved by this issue is that with the help of the new `Row Value Expression SPI`, you can configure ShardingSphere without using Groovy syntax at all through the buildArgs of `--report-unsupported-elements-at-runtim`, which makes  building a GraalVM Native Image that completes nativeTest becomes a reality.  

- If needed, I can help investigate the nativeTest failure issue on the `quarkus-shardingsphere` side, but this requires `quarkus-shardingsphere` to synchronize the `master` branch of shardingsphere.  In addition, the naming of related submodules may still be variable in shardingsphere.

**Comment by zhfeng** [comment]
Well, `quarkus-shardingsphere` does not depend on GraalVM Reachability Metadata directly but I think it could be helpful to build native images since `native-image` tool can detect the `.json` configuartion in the jars and add them automatically. Also I think we need to avoid using `--report-unsupported-elements-at-runtim` because it can lead to any unexpected error at runtime.

**Comment by linghengqian** [comment]
> Well, `quarkus-shardingsphere` does not depend on GraalVM Reachability Metadata directly but I think it could be helpful to build native images since `native-image` tool can detect the `.json` configuartion in the jars and add them automatically. Also I think we need to avoid using `--report-unsupported-elements-at-runtim` because it can lead to any unexpected error at runtime.

- This is natural. The current need to initialize GroovyShell related classes is because the default implementation of Row Value 
Expression SPI uses GroovyShell, which results in the need for downstream to actively add buildArgs for compatibility.  Maybe I can modify `shardingsphere-infra-expr-core` to expose a possible configuration item on the ShardingSphere side.  Or I can break this non-existent compatibility under GraalVM Native Image.

**Comment by zhfeng** [comment]
OK, thank for clarifying.

**Comment by linghengqian** [comment]
> Well, `quarkus-shardingsphere` does not depend on GraalVM Reachability Metadata directly but I think it could be helpful to build native images since `native-image` tool can detect the `.json` configuartion in the jars and add them automatically. Also I think we need to avoid using `--report-unsupported-elements-at-runtim` because it can lead to any unexpected error at runtime.

- @zhfeng I'll try to handle it at https://github.com/apache/shardingsphere/issues/29052 .
